include/lala/vnnlib_parser.hpp

```diff
@@ -0,0 +1,210 @@
+// Copyright 2025 Yi-Nung Tsao
+
+#ifndef LALA_PARSING_VNNLIB_PARSER_HPP
+#define LALA_PARSING_VNNLIB_PARSER_HPP
+
+#include <any>
+#include <cassert>
+#include <cstddef>
+#include <cstdlib>
+#include <fstream>
+#include <iostream>
+#include <iterator>
+#include <string>
+#include <system_error>
+
+#include "battery/shared_ptr.hpp"
+#include "lala/logic/ast.hpp"
+#include "peglib.h"
+
+
+namespace lala {
+
+namespace impl {
+
+template <class Allocator>
+class VnnlibParser {
+  using allocator_type = Allocator;
+  using F = TFormula<allocator_type>;
+  using SV = peg::SemanticValues;
+  using So = Sort<allocator_type>;
+  using FSeq = typename F::Sequence;
+
+  bool error;   // If an error was found during parsing.
+  bool silent;  // If we do not want to output error messages.
+  F &onnx_formulas;
+
+ public:
+  VnnlibParser(F &f) : error(false), silent(false), onnx_formulas(f) {}
+
+  battery::shared_ptr<F, allocator_type> parse(const std::string& input) {
+			peg::parser parser(R"(
+				Statements    <- (DeclareVar / Assertion / Comment)+
+
+				Integer       <- < [+-]?[0-9]+ >
+				Real          <- < ('inf' / '-inf' /
+														[+-]?[0-9]+ (('.' (&'..' / !'.') [0-9]*) /
+														([Ee][+-]?[0-9]+)) ) >
+				Identifier    <- < [a-zA-Z_][a-zA-Z0-9_]* >
+
+				BinaryOp      <- < '<=' / '>=' >
+				LogicOp       <- < 'and' / 'or' >
+
+				DeclareVar    <- '(' 'declare-const' Identifier 'Real' ')'
+				Bound         <- '(' BinaryOp Identifier (Real / Integer / Identifier) ')'
+				Constraint    <- '(' LogicOp Bound* ')'
+				Assertion     <- '(' 'assert' Bound ')' / '(' 'assert' '(' LogicOp+ Constraint* '))'
+
+				~Comment       <- ';' [^\n\r]* [ \n\r\t]*
+				%whitespace   <- [ \n\r\t]*
+			)");
+    assert(static_cast<bool>(parser) == true);
+
+    parser["Statements"] = [this](const SV& sv) { return make_statements(sv); };
+    parser["Integer"] = [](const SV& sv) { return F::make_z(sv.token_to_number<logic_int>()); };
+    parser["Real"] = [](const SV& sv) { return F::make_real(string_to_real(sv.token_to_string())); };
+    parser["Identifier"] = [](const SV& sv) { return sv.token_to_string(); };
+    parser["BinaryOp"] = [](const SV& sv) { return sv.token_to_string(); };
+    parser["LogicOp"] = [](const SV& sv) { return sv.token_to_string(); };
+    parser["DeclareVar"] = [this](const SV& sv) { return make_variable_decl(sv); };
+    parser["Bound"] = [this](const SV& sv) { return make_bound(sv); };
+    parser["Constraint"] = [this](const SV& sv) { return make_constraint(sv); };
+    parser["Assertion"] = [this](const SV& sv) { return make_assertion(sv); };
+
+    // std::cout << input.c_str() << std::endl;
+    FSeq seq;
+    F vnnlib_formulas;
+    if (parser.parse(input.c_str(), vnnlib_formulas) && !error) {
+      seq.push_back(onnx_formulas);
```

Comment:
please, separate VNN from ONNX, VNN shouldn't be linked to ONNX. And maybe we should rename VNN to smtlib?